### PR TITLE
Integrated query pushdown with Spark 3.1 DSV2 connector 

### PR DESCRIPTION
**Pull request by VinayLondhe14**

### CODE REVIEW

File: spark-bigquery-pushdown/pushdown_common_src/main/scala/com/google/cloud/spark/bigquery/pushdowns/SparkBigQueryPushdownUtil.scala [hunk 1]
```diff
@@ -126,4 +130,12 @@ object SparkBigQueryPushdownUtil {
       }
     }
   }
+
+  def getTableNameFromOptions(options: Map[String, String]): String = {
```
Comment (davidrabinowitz) [inline spark-bigquery-pushdown/pushdown_common_src/main/scala/com/google/cloud/spark/bigquery/pushdowns/SparkBigQueryPushdownUtil.scala:134]: Please check SparkBigQueryUtil.parseSimpleTaleId().

Also, as this method is not pushdown specific, it should be in SparkBigQueryUtil

File: spark-bigquery-pushdown/spark-3.1-bigquery-pushdown_2.12/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark31BigQueryStrategy.scala [hunk 2]
```diff
@@ -1,11 +1,31 @@
...
 
 class Spark31BigQueryStrategy(expressionConverter: SparkExpressionConverter, expressionFactory: SparkExpressionFactory, sparkPlanFactory: SparkPlanFactory)
   extends BigQueryStrategy(expressionConverter, expressionFactory, sparkPlanFactory) {
   override def generateQueryFromPlanForDataSourceV2(plan: LogicalPlan): Option[BigQuerySQLQuery] = {
-    throw new BigQueryPushdownUnsupportedException("Query pushdown for Spark 3.1 DataSourceV2 connector has not been implemented yet")
+    // DataSourceV2ScanRelation is the relation that is used in the Spark 3.1 DatasourceV2 connector
+    plan match {
+      case scanRelation@DataSourceV2ScanRelation(_, _, _) =>
+        // Get the scan and cast it to SupportsQueryPushdown to get the BigQueryRDDFactory
+        val scan = scanRelation.scan.asInstanceOf[SupportsQueryPushdown]
```
Comment (davidrabinowitz) [inline spark-bigquery-pushdown/spark-3.1-bigquery-pushdown_2.12/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark31BigQueryStrategy.scala:15]: What happens if a scan does not implement SupportsQueryPushdown ? A FileScan for example?

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-pushdown/spark-3.1-bigquery-pushdown_2.12/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark31BigQueryStrategy.scala (hunk 2 above)
Comment (VinayLondhe14) [reply spark-bigquery-pushdown/spark-3.1-bigquery-pushdown_2.12/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark31BigQueryStrategy.scala]: Good question. I updated the code to pattern match on `SupportsQueryPushdown`. 

If a scan does not implement SupportsQueryPushdown, then `None` will be returned from this method and we will resort to our regular non query pushdown code path

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**